### PR TITLE
Syscall and sysvar last restart slot

### PR DESCRIPTION
**Pull request by godmodegalactus**
A new SIMD to add new sysvar to get the last hard fork slot. Dapp use this information to get data about when cluster was restarted.

### CODE REVIEW

File: proposals/0047-last-hard-fork-syscall.md
Comment (mvines) [inline proposals/0047-last-hard-fork-syscall.md:15]: The sysvar can be skipped entirely, this only needs to be a syscall who's signature can be described here instead.

File: proposals/0047-last-hard-fork-syscall.md
Comment (mvines) [inline proposals/0047-last-hard-fork-syscall.md:21]: This definition of optimistically confirmed slot  is incorrect. I suggest removing the definition entirely since this SIMD is not introducing the term

File: proposals/0047-last-hard-fork-syscall.md
Comment (mvines) [inline proposals/0047-last-hard-fork-syscall.md:19]:
```suggestion
In Solana, when the cluster cannot reach consensus, it is currently restarted using
```

File: proposals/0047-last-hard-fork-syscall.md
Comment (mvines) [inline proposals/0047-last-hard-fork-syscall.md:29]: "After successfully restarting the network the hard fork slot is stored
in a variable called `hard_forks` in bank." -- remove. This is an implementation detail for the Solana Labs validator client, and may not be true for other clients

File: proposals/0047-last-hard-fork-syscall.md
Comment (mvines) [inline proposals/0047-last-hard-fork-syscall.md:51]: Let's not create a sysvar at all, this can be a pure syscall

**Review by mvines — COMMENTED** [review]
Thanks for doing this, idea looks all good

File: proposals/0047-last-hard-fork-syscall.md
Comment (godmodegalactus) [reply proposals/0047-last-hard-fork-syscall.md]: I have removed the variable name, but kept the implementation details. I specified that they are solana validator client.

I can remove that too but I do not see what I could write over here.

**Review by mvines — APPROVED** [review]
Some nits:
1. Please replace all usage of "dapp" in the proposal with "program". Dapp is a not a thing that's defined at the protocol level, if it even has a definition. But this syscall is very much used by (Solana) programs, so using this specific [term](https://docs.solana.com/terminology#program) helps to avoid confusion
2. Let's specify somewhere that Slot 0 is always considered to be a hard fork from the perspective of this syscall. Otherwise we need to deal with an `Option<Slot>` response, which is yuck.

I'm pretty ok with the meat here so soft approve from me. It'd be nice to get a couple more eyes on the text here before we call it done.

Some other possible reviewers: @pgarg66, @t-nelson, or @joncinque

**Review by t-nelson — COMMENTED** [review]
this looks great, thanks!  as a bit of a bike shed, since it seems like this will receive little contention, i wonder if we should give it the more generic name of "last restart slot".  motivation here being that there is a concurrent effort to improve/automate the restart process which, if completely successful, could one day eliminate hard forking as a requirement during many restart scenarios.  even if this eventuality comes true, there will be _some_ notion of a restart slot, whether a hardfork or w/e the improved restart solution may provide.  if we use a more generic name, we can change the backend implementation as needed without functionally effecting deployed programs. wdyt?

**Comment by godmodegalactus** [comment]
> this looks great, thanks! as a bit of a bike shed, since it seems like this will receive little contention, i wonder if we should give it the more generic name of "last restart slot". motivation here being that there is a concurrent effort to improve/automate the restart process which, if completely successful, could one day eliminate hard forking as a requirement during many restart scenarios. even if this eventuality comes true, there will be _some_ notion of a restart slot, whether a hardfork or w/e the improved restart solution may provide. if we use a more generic name, we can change the backend implementation as needed without functionally effecting deployed programs. wdyt?

I can update it to last restart slot instead of last hardfork in the title. In the motivation I can mention that currently
we have to hardfork to restart the network. So we consider hardfork slot as the restart slot. But may be if this evolves
then we can replace the implementation by returning a more appropriate measure instead of hardforks.

**Comment by godmodegalactus** [comment]
> Some nits:
> 
>     1. Please replace all usage of "dapp" in the proposal with "program". Dapp is a not a thing that's defined at the protocol level, if it even has a definition. But this syscall is very much used by (Solana) programs, so using this specific [term](https://docs.solana.com/terminology#program) helps to avoid confusion
> 
>     2. Let's specify somewhere that Slot 0 is always considered to be a hard fork from the perspective of this syscall. Otherwise we need to deal with an `Option<Slot>` response, which is yuck.
> 
> 
> I'm pretty ok with the meat here so soft approve from me. It'd be nice to get a couple more eyes on the text here before we call it done.
> 
> Some other possible reviewers: @pgarg66, @t-nelson, or @joncinque

Nit changes done.

File: proposals/0047-last-hard-fork-syscall.md
Comment (joncinque) [inline proposals/0047-last-hard-fork-syscall.md:84]: For frontends interacting with on-chain programs, they may also need to know the restart slot. For example, if the Mango program halts all settling 1000 slots after a hard fork, the frontend may like to show a message "no settling currently allowed, please wait X slots".

For frontends to fetch the information, we can expose the hard fork through an additional RPC call or sysvar account that exposes the entire `HardForks` structure. I don't feel too strongly either way, but I would lean towards the sysvar account since it gives more info and reduces RPC bloat.

On the other hand, given @t-nelson 's comment about generalizing this for future restart procedures, a new RPC call may be better. However, I don't know if we have enough information about the future restart process to know if the concept of a "restart slot" is even applicable.

Any thoughts here?

**Review by joncinque — COMMENTED** [review]
Great job! Just one point about exposing this off-chain

File: proposals/0047-last-hard-fork-syscall.md
Comment (mvines) [reply proposals/0047-last-hard-fork-syscall.md:84]: I feel like at this point we don't need an explicit RPC method or add a sysvar just for cleaner frontend handling of this condition. Outages aren't common so as long as the program protects itself a standard transaction simulation by the frontend will produce an (probably generic) error for the user in this exceptional case.

Comment (t-nelson) [reply proposals/0047-last-hard-fork-syscall.md:84]: they can already update the ui by querying whatever account is storing the corresponding "proceed from restart" signal/slot for their program. trying to do this with an explicit rpc method means they'd have to somehow track state in the client, which will be error-prone at best. the way i see this function being used is something like

```
    match instruction {
        PostOrder(order) => market.do_post_order(order),
        RemoveOrder(order) => market.do_remove_order(order),
        Match(order1, order2) => {
            if market.open_since_slot < sol_last_restart_slot() {
                return Err(MarketIsPostOnly);
            }
            market.do_match(order1, order2)
        },
        OpenMarket => market.do_open(clock.slot),
    }
```

**Comment by t-nelson** [comment]
> I can update it to last restart slot instead of last hardfork in the title. In the motivation I can mention that currently we have to hardfork to restart the network. So we consider hardfork slot as the restart slot. But may be if this evolves then we can replace the implementation by returning a more appropriate measure instead of hardforks.

yep, sounds good.  for the sake of initial implementation, the "last restart slot" will be the last hard for slot.  so long as what the syscall returns is a strictly monotonically increasing slot that actually signifies a network outage, it doesn't matter what the source slot is, and that's the goal of generalizing the naming

**Comment by mvines** [comment]
> yep, sounds good. for the sake of initial implementation, the "last restart slot" will be the last hard for slot. so long as what the syscall returns is a strictly monotonically increasing slot that actually signifies a network outage, it doesn't matter what the source slot is, and that's the goal of generalizing the naming

hopefully not monotonically, but otherwise 👍🏼

File: proposals/0047-last-hard-fork-syscall.md
Comment (mvines) [inline proposals/0047-last-hard-fork-syscall.md:63]:
```suggestion
The implementation of this syscall is pretty straitforward. In the Solana Labs validator client all
```

File: proposals/0047-last-hard-fork-syscall.md
Comment (mvines) [inline proposals/0047-last-hard-fork-syscall.md:68]:
```suggestion
For other validator clients, we have to get the last hard fork slot information and make
```

File: proposals/0047-last-hard-fork-syscall.md
Comment (mvines) [inline proposals/0047-last-hard-fork-syscall.md:51]:
```suggestion
Currently in Solana Labs validator client, `hard fork` slots are a good indicator when
```

**Review by mvines — APPROVED** [review]

**Comment by mvines** [comment]
happy to re-approve once ci is green

**Review by mvines — DISMISSED** [review]

**Comment by mvines** [comment]
(happy to merge this once we get one more approval)

**Review by joncinque — APPROVED** [review]
Looks good! r+ other comments, that is

File: proposals/0047-last-hard-fork-syscall.md
Comment (mvines) [reply proposals/0047-last-hard-fork-syscall.md:47]: > The additional complexity of a syscall is worthwhile if this hard fork info needs to be accessed frequently, as loading in another account makes a txn more expensive.

Right, adding a sysvar account takes up tx account space which is something that DeFi programs tend to struggle with, although perhaps less now with LUTs.

The idea here is to do something quick, minimal, and low risk for existing programs to enable some kind of incremental improvement before the next restart (🔥🔥 ).

Generally we'd expect this syscall to be used before every defi op to sanity check the current network state

**Comment by t-nelson** [comment]
I think we should deemphasize "hard fork" throughout, since it's now just
the restart slot source in the initial implementation.

I was hoping to suggest all the changes for you  in comments, but will be
traveling for the next several hours. Happy to do that once I'm back to
reliable internet, but fine if you beat me to it

File: proposals/0047-last-hard-fork-syscall.md
Comment (Lichtso) [reply proposals/0047-last-hard-fork-syscall.md:47]: I think we need a sysvar anyway because syscalls also use sysvars under the hood to serve the request. And yes, syscalls will all become instructions of built-in programs. Meaning, we will try to avoid constant functions which don't perform any calculations in the future and use sysvars instead.

File: proposals/0047-last-hard-fork-syscall.md
Comment (godmodegalactus) [reply proposals/0047-last-hard-fork-syscall.md:90]: I think the signature of the new syscall should match between clients.

Comment (godmodegalactus) [reply proposals/0047-last-hard-fork-syscall.md:90]: I am not sure what to add as test vectors.

File: proposals/0047-last-hard-fork-syscall.md
Comment (godmodegalactus) [reply proposals/0047-last-hard-fork-syscall.md:47]: I agree with mvines on this because with syscall we need not need to change the existing instruction signature to use this feature. Sysvar we have to update the program to pass the new sysvar inorder to use this feature. Already the limited tx account space is an issue. Syscall will really simplify things and serves the purpose for this feature.

File: proposals/0047-last-hard-fork-syscall.md
Comment (godmodegalactus) [reply proposals/0047-last-hard-fork-syscall.md]: Tried to fix most of them.

File: proposals/0047-last-hard-fork-syscall.md
Comment (mvines) [reply proposals/0047-last-hard-fork-syscall.md]: ok, thanks. Restoring the sysvar seems fine then, and that resolves the concern that Jon had about front-end accessibility

Comment (godmodegalactus) [reply proposals/0047-last-hard-fork-syscall.md]: Could we have both syscall and sysvar?

File: proposals/0047-last-hard-fork-syscall.md
Comment (godmodegalactus) [reply proposals/0047-last-hard-fork-syscall.md]: @mvines @t-nelson what do you guys think?

I have provided syscall name, I can add syscall id (murmur3 hash). 

I am correct we should not any code or pseudo code to the SIMD.

File: proposals/0047-last-hard-fork-syscall.md
Comment (gauravjaincr7) [reply proposals/0047-last-hard-fork-syscall.md]: outdated

File: proposals/0047-last-hard-fork-syscall.md
Comment (gauravjaincr7) [reply proposals/0047-last-hard-fork-syscall.md]: ++,Is it possible for us to incorporate both syscall and sysvar?

File: proposals/0047-last-hard-fork-syscall.md
Comment (gauravjaincr7) [reply proposals/0047-last-hard-fork-syscall.md]: so this implies the implementation section appears to be repetitive?? The SIMD should clearly outline the runtime modifications and provide some sample test vectors. However, it is at the discretion of each client to determine how to incorporate them

**Comment by gauravjaincr7** [comment]
check my comments

File: proposals/0047-last-hard-fork-syscall.md
Comment (godmodegalactus) [reply proposals/0047-last-hard-fork-syscall.md]: Ok understood thanks for the details.

File: proposals/0047-syscall-to-get-last-restart-slot.md
Comment (mschneider) [inline proposals/0047-syscall-to-get-last-restart-slot.md:50]: block time vs slot increase: when cluster time jumps by more than 1m between two slots, there was most likely a restart a few slots ago. issue is that this change is slightly delayed as it needs a regular stake majority to vote after the restart.

Comment (mschneider) [reply proposals/0047-syscall-to-get-last-restart-slot.md:50]: Here is an example from a recent restart:

403 was the fork
404 was still a vob
405 had user transactions
406 the cluster clock was updated

File: proposals/0047-last-hard-fork-syscall.md
Comment (mschneider) [reply proposals/0047-last-hard-fork-syscall.md]: it's very annoying to retro-actively add a sysvar to a running program, but if the syscall is anyways going away with the new runtime and won't be accessible anymore (if i understood @Lichtso correctly) let's rip that band-aid off now, rather than later no?

File: proposals/0047-last-hard-fork-syscall.md
Comment (godmodegalactus) [reply proposals/0047-last-hard-fork-syscall.md]: > @godmodegalactus @gauravjaincr7 I guess pseudocode is optional here since the algorithm is trivial (it just returns some externally-set value). My original point was that SIMDs should be implementation-agnostic (i.e. not refer to any Rust source files of a specific client). Because otherwise this SIMD only makes sense to devs who are familiar with that particular implementation. Things like `definitions.rs` might also be renamed in the future. Instead, I suggest describing the actual changes to the protocol.
> 
> Which seems to be:
> 
>     1. Feature gate: Since this change is a hardfork itself, protect it with a feature gate
> 
>     2. eBPF instruction set (verifier now accepts one more immediate value for the call instruction, i.e. the syscall ID); New builtin syscall handler and compute unit cost component
> 
>     3. Snapshot format: Changing the bank data structure might introduce an additional field to the snapshot protobuf, but I'm not sure if that's the case here
> 
>     4. Bank hash: If `solana-validator` is restarting from a snapshot from social consensus, the bank hash needs to change to indicate a hardfork
> 
> 
> Something like
> 
> ```
> ### Feature Gate
> 
> The runtime changes made by this proposal are activated by a feature gate with the address:
> 
>     <ADDRESS>
> 
> Upon activation, the functionality described below immediately becomes available.
> 
> ### New syscall
> 
> Newly deployed and upgraded programs may invoke syscall `sol_get_last_restart_slot` (syscall ID <0x...>).
> Each invocation consumes the syscall base amount of compute units (`<number>` on Solana mainnet).
> 
> Register `r0` is set to the last restart slot number (an unsigned 64-bit integer). All other GPRs are not modified.
> 
> Pseudocode:
> 
> # consumes given amount of compute units and may abort execution if compute units are exceeded
> fn consume_cpu(amount)
> 
> # Syscall handler
> fn sol_get_last_restart_slot():
>     consume_cu(SYSCALL_BASE_AMOUNT)
>     r0 = last_restart_slot
> 
> The syscall will be available in architectures eBPF and SBF. (TODO: Ask @Lichtso if this should also be available in SBFv2)
> ```

Thanks for details. By the way it raises a question, do we really need to have a feature gate for this feature?

**Review by gauravjaincr7 — COMMENTED** [review]
reviewed

**Review by gauravjaincr7 — CHANGES_REQUESTED** [review]
reviewed

File: proposals/0047-last-hard-fork-syscall.md
Comment (Lichtso) [reply proposals/0047-last-hard-fork-syscall.md]: > Could we have both syscall and sysvar?

Yes, that is probably the way to go.
Though, the syscall will only exist temporarily in the current program runtime and will be deprecated with it.

File: proposals/0047-last-hard-fork-syscall.md
Comment (godmodegalactus) [reply proposals/0047-last-hard-fork-syscall.md]: Yeah I understand, yes this is highly likely if this feature is not added without any feature gate.

Ok noted I will add it too.

File: proposals/0047-last-hard-fork-syscall.md
Comment (godmodegalactus) [reply proposals/0047-last-hard-fork-syscall.md]: I will update the SIMD accordingly then. Thanks.

Comment (godmodegalactus) [reply proposals/0047-last-hard-fork-syscall.md]: Done

File: proposals/0047-last-hard-fork-syscall.md
Comment (godmodegalactus) [reply proposals/0047-last-hard-fork-syscall.md]: Added some more information as requested.

File: proposals/0047-syscall-and-sysvar-for-last-restart-slot.md
Comment (hydrogenbond007) [inline proposals/0047-syscall-and-sysvar-for-last-restart-slot.md:27]: isnt stake a dynamic unit, what if some validators plan to drop staking and the new stake is less than 70% of the older stake

Comment (godmodegalactus) [reply proposals/0047-syscall-and-sysvar-for-last-restart-slot.md:27]: There are also some CLI arguments which help to disable stakes on nodes which do not participate in restart effort.

Comment (hydrogenbond007) [reply proposals/0047-syscall-and-sysvar-for-last-restart-slot.md:27]: @ripatel-jump @godmodegalactus thanks gotcha, was just curious

**Review by gauravjaincr7 — COMMENTED** [review]
reviewed

File: proposals/0047-syscall-and-sysvar-for-last-restart-slot.md
Comment (mvines) [reply proposals/0047-syscall-and-sysvar-for-last-restart-slot.md:58]: There's no change to the current restart procedure required by this SIMD actually

File: proposals/0047-syscall-and-sysvar-for-last-restart-slot.md
Comment (godmodegalactus) [reply proposals/0047-syscall-and-sysvar-for-last-restart-slot.md:127]: Yeah these numbers are somewhere in the solana validator code. There are some constants specified for each program.

File: proposals/0047-syscall-and-sysvar-for-last-restart-slot.md
Comment (godmodegalactus) [reply proposals/0047-syscall-and-sysvar-for-last-restart-slot.md:56]: I think it is not a new terminology per-se

File: proposals/0047-syscall-and-sysvar-for-last-restart-slot.md
Comment (IntokuSatori) [inline proposals/0047-syscall-and-sysvar-for-last-restart-slot.md:83]: In the next EPOCH.

**Review by gauravjaincr7 — APPROVED** [review]
reviewed

**Review by gauravjaincr7 — APPROVED** [review]
all the changes are valid and up to the point

File: proposals/0047-syscall-and-sysvar-for-last-restart-slot.md
Comment (gauravjaincr7) [reply proposals/0047-syscall-and-sysvar-for-last-restart-slot.md:127]: It is possible to define a single fixed CU price for each invocation of the syscall. However, the compute budget is a measure of the overall computational resources that are allocated to a particular process or system. In order to determine the appropriate compute budget for a given system, it is important to take into account various factors such as the complexity of the system, the number of users, and the expected usage patterns.

Therefore, it may be beneficial to have a more nuanced approach to defining the compute budget for the syscall, taking into account these various factors. However, as you mentioned, since this doc is still a draft, it's acceptable to use an arbitrary number for now and revise it later based on further analysis and testing. It's also a good idea to consult the CU defaults to get a baseline understanding of the expected CU consumption for the syscall.

**Review by gauravjaincr7 — APPROVED** [review]

**Review by gauravjaincr7 — COMMENTED** [review]
checking

File: proposals/0047-syscall-and-sysvar-for-last-restart-slot.md
Comment (godmodegalactus) [reply proposals/0047-syscall-and-sysvar-for-last-restart-slot.md]: Changes done

File: proposals/0047-syscall-and-sysvar-for-last-restart-slot.md
Comment (godmodegalactus) [reply proposals/0047-syscall-and-sysvar-for-last-restart-slot.md]: changes done.

File: proposals/0047-syscall-and-sysvar-for-last-restart-slot.md
Comment (godmodegalactus) [reply proposals/0047-syscall-and-sysvar-for-last-restart-slot.md]: changes done.

File: proposals/0047-syscall-and-sysvar-for-last-restart-slot.md
Comment (godmodegalactus) [reply proposals/0047-syscall-and-sysvar-for-last-restart-slot.md]: changes done.

**Review by jacobcreech — APPROVED** [review]
Seems like we've reached general consensus. Thank you for you contribution @godmodegalactus!

**Comment by mschneider** [comment]
@ripatel-fd how far along is the FD implementation of this syscall? the feature on anza / jito side is ready for testnet testing as far as i know.

**Pull request merged**
